sdk/python/feast/staging/storage_client.py

```diff
@@ -0,0 +1,312 @@
+
+        Args:
+            file_uri (urllib.parse.ParseResult): Parsed uri of the file ex: urlparse("gs://bucket/file.avro")
+        Returns:
+            TemporaryFile object
+        """
+        protocol = self._get_staging_protocol(file_uri.scheme)
+        return protocol.download_file(file_uri)
+
+    def execute_get_source_files(self, source: str) -> List[str]:
+        """
+        Lists all the files under a directory if path has wildcard(*) character.
+
+        Args:
+            source (str): File path with the protocol ex: gs://bucket/* or gs://bucket/file.avro
+        Returns:
+            List[str]: A list containing the full path to the file(s) in the remote staging location.
+        """
+        uri = urlparse(source)
+        if "*" in uri.path:
+            protocol = self._get_staging_protocol(uri.scheme)
+            return protocol.list_files(bucket=uri.hostname, uri=uri)
+        elif PROTOCOL(uri.scheme) in [PROTOCOL.S3, PROTOCOL.GS]:
+            return [source]
+        else:
+            raise Exception(
+                f"Could not identify file protocol {uri.scheme}. Only gs:// and file:// and s3:// supported"
+            )
+
+    def execute_file_upload(
+        self, scheme: str, local_path: str, bucket: str, remote_path: str
+    ):
+        """
+        Uploads file to a cloud storage, currently s3 and gs are supported
+
+        Args:
+            scheme (str): uri scheme: s3 or gs
+            local_path (str): Path to the local file that needs to be uploaded/staged
+            bucket (str): s3 or gs Bucket name
+            remote_path (str): relative path to the folder to which the files need to be uploaded
+        """
+        protocol = self._get_staging_protocol(scheme)
+        return protocol.upload_file(local_path, bucket, remote_path)
+
+    def _get_staging_protocol(self, protocol):
+        """
+        Lazy initialization of a specific protocol object(GCSProtocol, S3Protocol etc.) w
+        hen the 1st instance is encountered.
+
+        Args:
+            protocol (str): uri scheme: s3, gs or file
+
+        Returns:
+            An object of concrete implementation of AbstractStagingProtocol
+        """
+        if protocol in self._protocol_dict:
+            return self._protocol_dict[protocol]
+        else:
+            if PROTOCOL(protocol) == PROTOCOL.GS:
+                self._protocol_dict[protocol] = GCSProtocol()
+            elif PROTOCOL(protocol) == PROTOCOL.S3:
+                self._protocol_dict[protocol] = S3Protocol()
+            elif PROTOCOL(protocol) == PROTOCOL.LOCAL_FILE:
+                self._protocol_dict[protocol] = LocalFSProtocol()
+            else:
+                raise Exception(
+                    f"Could not identify file protocol {protocol}. Only gs:// and file:// and s3:// supported"
+                )
+            return self._protocol_dict[protocol]
+
+
+class AbstractStagingProtocol(ABC):
```

Comment:
Not sure about naming this `Protocol` when it clearly defines methods that relate storage client methods. Especially since we already have `PROTOCOL` enum, it might cause confusion.

Reply:
@mrzzy `AbstractStagingClient` would be appropriate?

Reply:
I think storage client/backend might be more suitable as generic list/upload/download methods (we expose here don't really have anything to do with staging even though we use it for staging